## Gather

In [1]:
import pandas as pd

In [2]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

## Assess

In [3]:
patients

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state  zip_code  \
0         576 Brown Bear Drive  Rancho California  California   92390.0   
1    2370 University Hill Road          Armstrong    Illinois   61812.0   
2        1493 Poling Farm Road               York    Nebraska   68467.0   
3          2335 Webster Street         Woodbridge          NJ    7095.0   
4         1428 Turkey Pen Lane             Dothan          AL   36303.0   
..                         ...                ...         ...       ...   
498        2530 Victoria Court       Milton Mills          ME    3852.0   
499   494 Clarksburg Park Road             Sedona          AZ   86341.0   
500          649 Nutter Street      Overland Park          MO   64110.0   
501   3652 Boone Crockett Lane            Seattle          WA   98109.0   
502          2778 North Avenue               Burr    Nebraska   68324.0   

           country                                          contact  \
0    United States             951-719-9170ZoeWellish@superrito.com   
1    United States            PamelaSHill@cuvox.de+1 (217) 569-3204   
2    United States                 402-363-6804JaeMDebord@gustr.com   
3    United States       PhanBaLiem@jourrapide.com+1 (732) 636-8246   
4    United States                  334-515-7487TimNeudorf@cuvox.de   
..             ...                                              ...   
498  United States      207-477-0579MustafaLindstrom@jourrapide.com   
499  United States               928-284-4492RumanBisliev@gustr.com   
500  United States            816-223-6007JinkedeKeizer@teleworm.us   
501  United States  ChidaluOnyekaozulu@jourrapide.com1 360 443 2060   
502  United States             PatrickGersten@rhyta.com402-848-4923   

     birthdate  weight  height   bmi  
0    7/10/1976   121.7      66  19.6  
1     4/3/1967   118.8      66  19.2  
2    2/19/1980   177.8      71  24.8  
3    7/26/1951   220.9      70  31.7  
4    2/18/1928   192.3      27  26.1  
..         ...     ...     ...   ...  
498  4/10/1959   181.1      72  24.6  
499  3/26/1948   239.6      70  34.4  
500  1/13/1971   171.2      67  26.8  
501  2/13/1952   176.9      67  27.7  
502   5/3/1954   138.2      71  19.3  

[503 rows x 14 columns]

In [4]:
treatments

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika      jindrová  41u - 48u          -         7.63       7.20   
1       elliot    richardson          -  40u - 45u         7.56       7.09   
2     yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3         skye    gormanston  33u - 36u          -         7.97       7.62   
4       alissa        montez          -  33u - 29u         7.78       7.46   
..         ...           ...        ...        ...          ...        ...   
275     albina      zetticci  45u - 51u          -         7.93       7.73   
276       john   teichelmann          -  49u - 49u         7.90       7.58   
277     mathea       lillebø  23u - 36u          -         9.04       8.67   
278     vallie        prince  31u - 38u          -         7.64       7.28   
279     samúel  guðbrandsson  53u - 56u          -         8.00       7.64   

     hba1c_change  
0             NaN  
1            0.97  
2             NaN  
3            0.35  
4            0.32  
..            ...  
275          0.20  
276           NaN  
277          0.37  
278          0.36  
279          0.36  

[280 rows x 7 columns]

In [5]:
adverse_reactions

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

In [6]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [7]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [8]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


In [9]:
all_columns = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

In [10]:
list(patients)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'contact',
 'birthdate',
 'weight',
 'height',
 'bmi']

In [11]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [12]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [13]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [14]:
patients.sample(5)

patient_id assigned_sex given_name  surname                address  \
117         118         male        Sửu       Tạ        3464 Big Indian   
318         319         male     Benoît   Bonami     230 Lady Bug Drive   
402         403       female    Manouck  Wubbels  4789 Devils Hill Road   
470         471         male   Muhammad   Holden   3414 Franklin Avenue   
461         462         male     Cannan  Cabrera    2102 Geraldine Lane   

            city     state  zip_code        country  \
117  New Orleans        LA   70112.0  United States   
318  Garden City  New York   11530.0  United States   
402        Utica        MS   39175.0  United States   
470     Live Oak   Florida   32064.0  United States   
461     New York        NY   10014.0  United States   

                                            contact   birthdate  weight  \
117           TaQuiSuu@teleworm.us+1 (504) 528-8317   5/24/1928   182.2   
318            BenoitBonami@gustr.com1 718 954 8136   4/13/1975   162.8   
402          601-885-6550ManouckWubbels@armyspy.com   11/3/1964   201.5   
470          386-234-5932MuhammadHolden@fleckens.hu    9/8/1969   159.5   
461  646-289-4177CannanCabreraOrdonez@superrito.com  10/12/1980   209.7   

     height   bmi  
117      67  28.5  
318      70  23.4  
402      65  33.5  
470      66  25.7  
461      71  29.2

In [15]:
patients.surname.value_counts()

Doe              6
Taylor           3
Jakobsen         3
Cindrić          2
Parker           2
                ..
Miller           1
Vieira           1
Obinna           1
Wolfe            1
Traustadóttir    1
Name: surname, Length: 466, dtype: int64

In [16]:
patients.address.value_counts()

123 Main Street         6
2476 Fulton Street      2
2778 North Avenue       2
648 Old Dear Lane       2
4148 Callison Lane      1
                       ..
4852 Rose Avenue        1
2607 Water Street       1
3130 Jessie Street      1
36 Heather Sees Way     1
1493 Randolph Street    1
Name: address, Length: 483, dtype: int64

In [17]:
patients[((patients.address.duplicated()==True) & (patients.address.isnull()==False))]
patients[patients.address=='2476 Fulton Street']

patient_id assigned_sex given_name surname             address      city  \
131         132       female     Sandra  Taylor  2476 Fulton Street  Rainelle   
282         283       female      Sandy  Taylor  2476 Fulton Street  Rainelle   

    state  zip_code        country                               contact  \
131    WV   25962.0  United States  304-438-2648SandraCTaylor@dayrep.com   
282    WV   25962.0  United States  304-438-2648SandraCTaylor@dayrep.com   

      birthdate  weight  height   bmi  
131  10/23/1960   206.1      64  35.4  
282  10/23/1960   206.1      64  35.4

In [18]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
       ...  
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [19]:
weight_lbs = patients[patients.surname == 'Zaitseva'].weight * 2.20462
height_in = patients[patients.surname == 'Zaitseva'].height
bmi_check = 703 * weight_lbs / (height_in * height_in)
bmi_check

210    19.055827
dtype: float64

In [20]:
patients[patients.surname == 'Zaitseva'].bmi

210    19.1
Name: bmi, dtype: float64

In [21]:
sum(treatments.auralin.isnull())

0

In [22]:
sum(treatments.novodra.isnull())

0

#### Quality
##### `patients` table
- Zip code is a float not a string
- Zip code has four digits sometimes
- Tim Neudorf height is 27 in instead of 72 in
- Full state names sometimes, abbreviations other times
- Dsvid Gustafsson
- Missing demographic information (address - contact columns) ***(can't clean)***
- Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns)
- Multiple phone number formats
- Default John Doe data
- Multiple records for Jakobsen, Gersten, Taylor
- kgs instead of lbs for Zaitseva weight

##### `treatments` table
- Missing HbA1c changes
- The letter 'u' in starting and ending doses for Auralin and Novodra
- Lowercase given names and surnames
- Missing records (280 instead of 350)
- Erroneous datatypes (auralin and novodra columns)
- Inaccurate HbA1c changes (leading 4s mistaken as 9s)
- Nulls represented as dashes (-) in auralin and novodra columns

##### `adverse_reactions` table
- Lowercase given names and surnames

#### Tidiness
- Contact column in `patients` table should be split into phone number and email
- Three variables in two columns in `treatments` table (treatment, start dose and end dose)
- Adverse reaction should be part of the `treatments` table
- Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables

## Clean

In [23]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

### Missing Data

<font color='red'>Complete the following two "Missing Data" **Define, Code, and Test** sequences after watching the *"Address Missing Data First"* video.</font>

#### `treatments`: Missing records (280 instead of 350)

##### Define
- filling missing data in treatments with the data in treatments_cut
*Your definition here. Note: the missing `treatments` records are stored in a file named `treatments_cut.csv`, which you can see in this Jupyter Notebook's dashboard (click the **jupyter** logo in the top lefthand corner of this Notebook). Hint: [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.concat.html) for the function used in the solution.*

##### Code

In [24]:
# Your cleaning code here
treatments_cut = pd.read_csv('treatments_cut.csv')
treatments_cut.head()
treatments_clean = pd.concat([treatments_clean,treatments_cut],join='inner',
               ignore_index=True,keys=('given_name','surname'))


##### Test

In [25]:
treatments_clean.shape

(350, 7)

#### `treatments`: Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)
- re-calculating HbA1c changes by substracting hba1c_start from hba1c_end
*Note: the "Inaccurate HbA1c changes (leading 4s mistaken as 9s)" observation, which is an accuracy issue and not a completeness issue, is included in this header because it is also fixed by the cleaning operation that fixes the missing "Missing HbA1c changes" observation. Multiple observations in one **Define, Code, and Test** header occurs multiple times in this notebook.*

##### Define
*Your definition here.*

##### Code

In [26]:
# Your cleanitreatments_clean.hba1c_change = treatments_clean.hba1c_start-treatments_clean.hba1c_endng code here
treatments_clean.hba1c_change = treatments_clean.hba1c_start-treatments_clean.hba1c_end

##### Test

In [29]:
# Your testing code here
treatments_clean.head()


given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0          0.43  
1          0.47  
2          0.43  
3          0.35  
4          0.32

### Tidiness

<font color='red'>Complete the following four "Tidiness" **Define, Code, and Test** sequences after watching the *"Cleaning for Tidiness"* video.</font>

#### Contact column in `patients` table contains two variables: phone number and email

##### Define
- spreating contact column in patinets and make two new column phon and e-mail
*Your definition here. Hint 1: use regular expressions with pandas' [`str.extract` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.extract.html). Here is an amazing [regex tutorial](https://regexone.com/). Hint 2: [various phone number regex patterns](https://stackoverflow.com/questions/16699007/regular-expression-to-match-standard-10-digit-phone-number). Hint 3: [email address regex pattern](http://emailregex.com/), which you might need to modify to distinguish the email from the phone number.*

##### Code

In [30]:
# Your cleaning code here
patients_clean['phone-number']= patients_clean.contact.str.extract("((?:\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})")
patients_clean['email'] = patients_clean.contact.str.extract('([a-zA-Z][a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+[a-zA-Z])')
patients_clean = patients_clean.drop('contact', axis=1)

##### Test

In [34]:
# Your testing code here
patients_clean['email'], patients_clean['phone-number']


(0               ZoeWellish@superrito.com
 1                   PamelaSHill@cuvox.de
 2                   JaeMDebord@gustr.com
 3              PhanBaLiem@jourrapide.com
 4                    TimNeudorf@cuvox.de
                      ...                
 498      MustafaLindstrom@jourrapide.com
 499               RumanBisliev@gustr.com
 500            JinkedeKeizer@teleworm.us
 501    ChidaluOnyekaozulu@jourrapide.com
 502             PatrickGersten@rhyta.com
 Name: email, Length: 503, dtype: object,
 0           951-719-9170
 1      +1 (217) 569-3204
 2           402-363-6804
 3      +1 (732) 636-8246
 4           334-515-7487
              ...        
 498         207-477-0579
 499         928-284-4492
 500         816-223-6007
 501         360 443 2060
 502         402-848-4923
 Name: phone-number, Length: 503, dtype: object)

#### Three variables in two columns in `treatments` table (treatment, start dose and end dose)

##### Define
- making treatment column to include the type of treatment 
- sperating start and end dose for each treatment in seprates columns
*Your definition here. Hint: use pandas' [melt function](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html) and [`str.split()` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.split.html). Here is an excellent [`melt` tutorial](https://deparkes.co.uk/2016/10/28/reshape-pandas-data-with-melt/).*

##### Code

In [36]:
# Your cleaning code here
treatments_clean = pd.melt(treatments_clean, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],
                           var_name='treatment', value_name='dose')

In [44]:
treatments_clean= treatments_clean[treatments_clean.dose!='-']

In [52]:
treatments_clean['dose_start'],treatments_clean['dose_end']=treatments_clean.dose.str.split(' - ').str

<ipython-input-52-936c793c6cde>:1: FutureWarning: Columnar iteration over characters will be deprecated in future releases.
  treatments_clean['dose_start'],treatments_clean['dose_end']=treatments_clean.dose.str.split(' - ').str


In [55]:
treatments_clean.drop('dose',axis=1,inplace=True)

##### Test

In [56]:
# Your testing code here
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37          0.39   auralin   

  dose_start dose_end  
0        41u      48u  
3        33u      36u  
6        37u      42u  
7        31u      38u  
9        30u      36u

#### Adverse reaction should be part of the `treatments` table

##### Define
- add Adverse reaction to treatment table
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) for the function used in the solution.*

##### Code

In [59]:
# Your cleaning code here
treatments_clean= pd.merge(treatments_clean, adverse_reactions_clean,
                            on=['given_name', 'surname'], how='left')

##### Test

In [68]:
treatments_clean.shape

(350, 9)

#### Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables  and Lowercase given names and surnames

##### Define
- making surname and given name in lowercase and then join treatments to patient id and then drop given-name and surr name from it 
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) for one function used in the solution and [tutorial](http://erikrood.com/Python_References/dropping_rows_cols_pandas.html) for another function used in the solution.*

##### Code

In [83]:
# Your cleaning code here
patients_sub= patients_clean[['patient_id','given_name','surname']]
treatments_clean.given_name=treatments_clean.given_name.str.title()
treatments_clean.surname=treatments_clean.surname.str.title()
treatments_clean= pd.merge(patients_sub,treatments_clean,on=['given_name','surname'])

In [84]:
treatments_clean.drop(['given_name','surname'],axis=1,inplace=True)

##### Test

In [87]:
# Your testing code here
treatments_clean.shape

(344, 8)

### Quality

<font color='red'>Complete the remaining "Quality" **Define, Code, and Test** sequences after watching the *"Cleaning for Quality"* video.</font>

#### Zip code is a float not a string and Zip code has four digits sometimes

##### Define
- convert zip code column to string and add 0 at the begining if having 4 digits
*Your definition here. Hint: see the "Data Cleaning Process" page.*

##### Code

In [139]:
# Your cleaning code here
patients_clean.zip_code=patients_clean.zip_code.astype(str)
patients_clean.zip_cod= patients_clean.zip_code.str[:-2]
patients_clean.zip_cod=patients_clean.zip_cod.str.pad(5,fillchar='0')

##### Test

In [140]:
# Your testing code here
patients_clean.zip_cod.sample(5)

475    94041
139    45202
220    94612
17     30082
498    03852
Name: zip_code, dtype: object

#### Tim Neudorf height is 27 in instead of 72 in

##### Define
- change patient with hiegh 27 to be 72 in
*Your definition here.*

##### Code

In [142]:
# Your cleaning code here
patients_clean.height = patients_clean.height.replace(27,72) 

##### Test

In [148]:
# Your testing code here
patients_clean[patients_clean.patient_id==5]

patient_id assigned_sex given_name  surname               address    city  \
4           5         male        Tim  Neudorf  1428 Turkey Pen Lane  Dothan   

  state zip_code        country  birthdate  weight  height   bmi  \
4    AL  36303.0  United States  2/18/1928   192.3      72  26.1   

   phone-number                email  
4  334-515-7487  TimNeudorf@cuvox.de

#### Full state names sometimes, abbreviations other times

##### Define
- change state fullname to it's abbreviations
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_apply_operations_to_dataframes.html) for method used in solution.*

##### Code

In [153]:
# Your cleaning code here
state_abbrev = {'California': 'CA',
                'New York': 'NY',
                'Illinois': 'IL',
                'Florida': 'FL',
                'Nebraska': 'NE'}
def change_fullname(x):
    if x in state_abbrev:
        return state_abbrev[x]
    else:
        return x
    
patients_clean.state= patients_clean.state.map(change_fullname)

##### Test

In [157]:
# Your testing code here
patients_clean.state

0      CA
1      IL
2      NE
3      NJ
4      AL
       ..
498    ME
499    AZ
500    MO
501    WA
502    NE
Name: state, Length: 503, dtype: object

#### Dsvid Gustafsson

##### Define
- change Dsvid to David
*Your definition here.*

##### Code

In [158]:
# Your cleaning code here
patients_clean.given_name=patients_clean.given_name.replace('Dsvid','David')

##### Test

In [159]:
# Your testing code here
patients_clean[patients_clean.surname=='Gustafsson']

patient_id assigned_sex given_name     surname             address  \
8           9         male      David  Gustafsson  1790 Nutter Street   

          city state zip_code        country birthdate  weight  height   bmi  \
8  Kansas City    MO  64105.0  United States  3/6/1937   163.9      66  26.5   

   phone-number                        email  
8  816-265-9578  DavidGustafsson@armyspy.com

#### Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns) and Erroneous datatypes (auralin and novodra columns) and The letter 'u' in starting and ending doses for Auralin and Novodra

##### Define
*Your definition here. Hint: [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.astype.html) for one method used in solution, [documentation page](http://pandas.pydata.org/pandas-docs/version/0.20/generated/pandas.to_datetime.html) for one function used in the solution, and [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.strip.html) for another method used in the solution.*

##### Code

In [163]:
# Your cleaning code here
patients_clean.assigned_sex=patients_clean.assigned_sex.astype('category')
patients_clean.birthdate= pd.to_datetime(patients_clean.birthdate)


In [167]:
treatments_clean.dose_start= treatments_clean.dose_start.str[:-1]
treatments_clean.dose_end= treatments_clean.dose_end.str[:-1]

In [168]:
treatments_clean.dose_start= treatments_clean.dose_start.astype(int)
treatments_clean.dose_end= treatments_clean.dose_end.astype(int)

##### Test

In [169]:
# Your testing code here
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   patient_id    503 non-null    int64         
 1   assigned_sex  503 non-null    category      
 2   given_name    503 non-null    object        
 3   surname       503 non-null    object        
 4   address       491 non-null    object        
 5   city          491 non-null    object        
 6   state         491 non-null    object        
 7   zip_code      503 non-null    object        
 8   country       491 non-null    object        
 9   birthdate     503 non-null    datetime64[ns]
 10  weight        503 non-null    float64       
 11  height        503 non-null    int64         
 12  bmi           503 non-null    float64       
 13  phone-number  491 non-null    object        
 14  email         491 non-null    object        
dtypes: category(1), datetime64[ns](1), float

In [170]:
treatments_clean.head()

patient_id  hba1c_start  hba1c_end  hba1c_change treatment  dose_start  \
0           1         7.71       7.30          0.41   novodra          33   
1           2         9.53       9.10          0.43   novodra          27   
2           4         7.58       7.10          0.48   novodra          43   
3           6         7.73       7.34          0.39   auralin          50   
4           7         7.65       7.26          0.39   novodra          32   

   dose_end adverse_reaction  
0        33              NaN  
1        29              NaN  
2        48              NaN  
3        60              NaN  
4        33              NaN

#### Multiple phone number formats

##### Define
- reomve any non-digit and 1 to the start of it (if <11)
*Your definition here. Hint: helpful [Stack Overflow answer](https://stackoverflow.com/a/123681).*

##### Code

In [172]:
# Your cleaning code here
patients_clean['phone-number']= patients_clean['phone-number'].str.replace(r'\D+','')
patients_clean['phone-number']= patients_clean['phone-number'].str.pad(11, fillchar='1')

<ipython-input-172-56e772bd547c>:2: FutureWarning: The default value of regex will change from True to False in a future version.
  patients_clean['phone-number']= patients_clean['phone-number'].str.replace(r'\D+','')


##### Test

In [173]:
# Your testing code here
patients_clean['phone-number']

0      19517199170
1      12175693204
2      14023636804
3      17326368246
4      13345157487
          ...     
498    12074770579
499    19282844492
500    18162236007
501    13604432060
502    14028484923
Name: phone-number, Length: 503, dtype: object

#### Default John Doe data

##### Define
- reomve Doe
*Your definition here. Recall that it is assumed that the data that this John Doe data displaced is not recoverable.*

##### Code

In [174]:
# Your cleaning code here
patients_clean = patients_clean[patients_clean.surname != 'Doe']

##### Test

In [179]:
# Your testing code here
patients_clean.surname.value_counts()

Taylor           3
Jakobsen         3
Kowalczyk        2
Silva            2
Lâm              2
                ..
Miller           1
Vieira           1
Obinna           1
Wolfe            1
Traustadóttir    1
Name: surname, Length: 465, dtype: int64

#### Multiple records for Jakobsen, Gersten, Taylor

##### Define
- romove only the column with surname Jakobsen, Gersten, Taylor that not affect treatments table
*Your definition here.*

##### Code

In [209]:
# Your cleaning code here
patients_clean= patients_clean[~((patients_clean.address.duplicated()) & patients_clean.address.notnull())]

##### Test

In [210]:
# Your testing code here
patients_clean[patients_clean.surname == 'Jakobsen']

patient_id assigned_sex given_name   surname             address  \
24           25         male      Jakob  Jakobsen   648 Old Dear Lane   
432         433       female      Karen  Jakobsen  1690 Fannie Street   

            city state zip_code        country  birthdate  weight  height  \
24   Port Jervis    NY  12771.0  United States 1985-08-01   155.8      67   
432      Houston    TX  77020.0  United States 1962-11-25   185.2      67   

      bmi phone-number                         email  
24   24.4  18458587707     JakobCJakobsen@einrot.com  
432  29.0  19792030438  KarenJakobsen@jourrapide.com

#### kgs instead of lbs for Zaitseva weight

##### Define
*Your definition here.*

##### Code

In [212]:
# Your cleaning code here
weight_kg = patients_clean.weight.min()
patients_clean.weight = patients_clean.weight.replace(weight_kg,weight_kg*2.20462)

C:\Users\kawrshy\anaconda3\lib\site-packages\pandas\core\generic.py:5494: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


##### Test

In [215]:
# Your testing code here
patients_clean[patients_clean.surname=='Zaitseva']

patient_id assigned_sex given_name   surname              address  \
210         211       female    Camilla  Zaitseva  4689 Briarhill Lane   

        city state zip_code        country  birthdate      weight  height  \
210  Wooster    OH  44691.0  United States 1938-11-26  107.585456      63   

      bmi phone-number                          email  
210  19.1  13302022145  CamillaZaitseva@superrito.com

In [223]:
treatments_clean.groupby(['treatment']).hba1c_change.mean()

treatment
auralin    0.388218
novodra    0.405176
Name: hba1c_change, dtype: float64

## Saving cleaned data

In [ ]:
treatments_clean.to_csv('treatments_clean.csv')
patients_clean.to_csv('patients_clean.csv')